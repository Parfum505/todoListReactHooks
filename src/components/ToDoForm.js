import React, { useState } from "react";

function ToDoForm({ addTask, lastId }) {
  const [todo, setTodo] = useState({
    id: 0,
    item: "",
    completed: false,
  });
  const [isValidatedForm, setIsValidatedForm] = useState("");
  function inputChange(e) {
    setTodo({ ...todo, item: e.target.value, id: lastId });
  }
  function handleSubmitForm(e) {
    e.preventDefault();
    if (!todo.item.trim()) {
      setIsValidatedForm("was-validated");
    } else {
      const newId = todo.id + 1;
      addTask(newId, todo);
      setTodo({ ...todo, item: "" });
      setIsValidatedForm("");
    }
  }
  return (
    <div className="row justify-content-center">
      <form
        data-testid="form"
        className={`col-11 col-sm-10 col-md-7 col-lg-5 mb-5 needs-validation ${isValidatedForm}`}
        onSubmit={handleSubmitForm}
        action=""
        noValidate
      >
        <div className="input-group input-group-lg">
          <input
            type="text"
            name="newTask"
            value={todo.item}
            className="form-control"
            placeholder="Enter text ..."
            id="todo"
            required
            onChange={inputChange}
          />
          <div className="input-group-append">
            <button id="submit_btn" type="submit" className="btn btn-primary">
              Add
            </button>
          </div>
          {isValidatedForm ? (
            <div className="invalid-feedback">Please fill out this field.</div>
          ) : null}
        </div>
      </form>
    </div>
  );
}

export default ToDoForm;
