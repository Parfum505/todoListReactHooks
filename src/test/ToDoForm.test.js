import React from "react";
import ToDoForm from "../components/ToDoForm";
import { render, fireEvent, cleanup } from "@testing-library/react";

afterEach(cleanup);

describe("ToDoForm component", () => {
  test("matches snapshot", () => {
    const { asFragment } = render(<ToDoForm />);
    expect(asFragment()).toMatchSnapshot();
  });
  test("form renders correctly", () => {
    const {
      getByPlaceholderText,
      getByText,
      getByTestId,
      queryByText,
    } = render(<ToDoForm />);
    expect(getByPlaceholderText("Enter text ...")).toBeInTheDocument();
    expect(getByText("Add")).toBeInTheDocument();
    expect(getByTestId("form")).toBeInTheDocument();
    expect(queryByText("Please fill out this field.")).toBeNull();
  });

  test("form submit correctly", () => {
    const mocAddTask = jest.fn();
    const { queryByText, getByPlaceholderText } = render(
      <ToDoForm addTask={mocAddTask} />
    );
    const input = getByPlaceholderText("Enter text ...");
    fireEvent.change(input, { target: { value: "test" } });
    fireEvent.click(queryByText("Add"));
    expect(queryByText("Please fill out this field.")).toBeNull();
  });

  test("form submit empty input", () => {
    const mocAddTask = jest.fn();
    const { queryByText, getByPlaceholderText } = render(
      <ToDoForm addTask={mocAddTask} />
    );
    const input = getByPlaceholderText("Enter text ...");
    fireEvent.change(input, { target: { value: "" } });
    fireEvent.click(queryByText("Add"));
    expect(queryByText("Please fill out this field.")).toBeInTheDocument();
  });

  test("input state updates on change", () => {
    const { queryByPlaceholderText } = render(<ToDoForm />);
    const input = queryByPlaceholderText("Enter text ...");
    fireEvent.change(input, { target: { value: "some text" } });
    expect(input.value).toBe("some text");
  });
});
