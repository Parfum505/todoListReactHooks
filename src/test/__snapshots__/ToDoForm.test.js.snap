// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDoForm component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="row justify-content-center"
  >
    <form
      action=""
      class="col-11 col-sm-10 col-md-7 col-lg-5 mb-5 needs-validation "
      data-testid="form"
      novalidate=""
    >
      <div
        class="input-group input-group-lg"
      >
        <input
          class="form-control"
          id="todo"
          name="newTask"
          placeholder="Enter text ..."
          required=""
          type="text"
          value=""
        />
        <div
          class="input-group-append"
        >
          <button
            class="btn btn-primary"
            id="submit_btn"
            type="submit"
          >
            Add
          </button>
        </div>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
